src/main/java/org/apache/mesos/scheduler/repair/RepairScheduler.java

```
@@ -0,0 +1,163 @@
+            }
+        } else {
+            log.info("Launching a stopped task");
+            replacingStopped = true;
+            offerReq = offerReqProvider.maybeGetNewOfferRequirement(targetConfigName, block);
+        }
+
+        if (offerReq.isPresent() && launchConstrainer.canLaunch(offerReq.get())) {
+            log.info("Preparing to launch task");
+            OfferEvaluator offerEvaluator = new OfferEvaluator();
+            List<OfferRecommendation> recommendations = offerEvaluator.evaluate(offerReq.get(), offers);
+            if (!recommendations.isEmpty()) {
+                // we're replacing some task(s)
+                Set<String> relaunchedTasks = recommendations.stream()
+                        .filter(it -> it.getOperation().hasLaunch())
+                        .flatMap(it -> it.getOperation().getLaunch().getTaskInfosList().stream())
+                        .map(TaskInfo::getName)
+                        .collect(Collectors.toSet());
+                if (replacingStopped) {
+                    // we'll filter their names out of the stopped pool
+                    newStopped = newStopped.stream()
+                            .filter(it -> !relaunchedTasks.contains(it))
+                            .collect(Collectors.toList());
+                } else {
+                    // we'll filter their names out of the failed pool
+                    newFailed = newFailed.stream()
+                            .filter(it -> !relaunchedTasks.contains(it))
+                            .collect(Collectors.toList());
+                }
+            }
+            List<Operation> launchOperations = recommendations.stream()
+                    .map(OfferRecommendation::getOperation)
+                    .filter(Operation::hasLaunch)
+                    .collect(Collectors.toList());
+            if (launchOperations.size() > 1) {
+                throw new RuntimeException("Repairs should only launch one task at a time, found "
```

A `RuntimeException` seems harsh here.  What is the expectation?  Would you like the Scheduler to crash on this unexpected scenario?  I would prefer to raise an Exception which requires declaration in the method signature such that consumers are forced to decide what do when encountering a Recovery system failure.